src/session.ts

```diff
@@ -1,29 +1,77 @@
 import { Context } from './context'
 import { Middleware } from './types'
 
-export function session<SessionData extends object>({
-  ttl = Infinity,
-  store = new Map<string, { expires: number; session: SessionData }>(),
-  getSessionKey = (ctx: Context) =>
-    ctx.from && ctx.chat && `${ctx.from.id}:${ctx.chat.id}`,
-} = {}): Middleware.ExtFn<Context, { session?: SessionData }> {
-  const ttlMs = ttl * 1000
+export type StorageKeyFn<T extends Context> = (
+  ctx: T
+) => Promise<string | undefined>
 
+export interface Storage<T> {
+  getItem: (name: string) => Promise<T | undefined>
+  setItem: (name: string, value: T) => Promise<void>
+  deleteItem: (name: string) => Promise<void>
+}
+
+/**
+ * session middleware
+ *
+ * more 3rd-party session middlware
+ *
+ * see https://npm.im/search?q=telegraf-session
+ */
+export function session<SessionData extends object>(
+  makeKey: StorageKeyFn<Context> = makeDefaultKey,
+  storage: Storage<SessionData> = new MemorySessionStorage<SessionData>()
+): Middleware.ExtFn<Context, { session?: SessionData }> {
```

Comment:
This ordering is definitely better than `storage, makeKey`. However, I think it would make sense to stick with the object version we've had before. That way, every user can choose what's best for them.